src/context.rs

```diff
@@ -1258,17 +1285,133 @@ impl SeafowlContext for DefaultSeafowlContext {
+                                let partitions = self
+                                    .partition_catalog
+                                    .load_table_partitions(table.table_version_id)
+                                    .await?;
+
+                                // To simulate the effect of a WHERE clause from a DELETE, we
+                                // need to use the inverse clause in a SELECT when filtering
+                                let filter = create_physical_expr(
+                                    &expr.clone().not(),
+                                    &table.schema.arrow_schema.clone().to_dfschema()?,
+                                    table.schema().as_ref(),
+                                    &ExecutionProps::new(),
+                                )?;
+
+                                match SeafowlPruningStatistics::from_partitions(
+                                    partitions.clone(),
+                                    table.schema(),
+                                ) {
+                                    Ok(pruning_stats) => {
+                                        // Determine the set of all partition ids that will need to
+                                        // be filtered
+                                        let partitions_to_filter =
+                                            HashSet::<PhysicalPartitionId>::from_iter(
+                                                pruning_stats
+                                                    .prune(&[expr.clone()])
+                                                    .await
+                                                    .iter()
+                                                    .map(|p| p.partition_id.unwrap()),
+                                            );
+
+                                        let mut final_partition_ids =
+                                            Vec::with_capacity(partitions.len());
+
+                                        // Group adjacent partitions eligible for filtering
+                                        let mut partitions_grouped: Vec<(
+                                            bool,
+                                            Vec<SeafowlPartition>,
+                                        )> = Vec::new();
+                                        for (keep, group) in
+                                            &partitions.into_iter().group_by(|p| {
+                                                !partitions_to_filter
+                                                    .contains(&p.partition_id.unwrap())
+                                            })
+                                        {
+                                            partitions_grouped
+                                                .push((keep, group.collect()))
+                                        }
+
+                                        for (keep, group) in partitions_grouped {
```

Comment:
The code doesn't use `partitions_grouped` a second time after iterating through the `group_by` result, so you can iterate through this immediately without first persisting in `partitions_grouped`:

```suggestion
                                        let mut final_partition_ids =
                                            Vec::with_capacity(partitions.len());

                                        // Group adjacent partitions eligible for filtering
                                        for (keep, group) in
                                            &partitions.into_iter().group_by(|p| {
                                                !partitions_to_filter
                                                    .contains(&p.partition_id.unwrap())
                                            })
                                        {
```

Reply:
Actually I tried that first, but I couldn't figure out how to appease the compiler; it looks like awaits can't be used inside `group_by`'s:
```rust
error: future cannot be sent between threads safely
    --> src/context.rs:1099:41
     |
1099 |       ) -> Result<Arc<dyn ExecutionPlan>> {
     |  _________________________________________^
1100 | |         // Similarly to DataFrame::sql, run certain logical plans outside of the actual execution flow
1101 | |         // and produce a dummy physical plan instead
1102 | |         match plan {
...    |
1505 | |         }
1506 | |     }
     | |_____^ future created by async block is not `Send`
     |
     = help: within `itertools::GroupBy<bool, std::vec::IntoIter<SeafowlPartition>, [closure@src/context.rs:1339:78: 1342:46]>`, the trait `std::marker::Sync` is not implemented for `RefCell<itertools::groupbylazy::GroupInner<bool, std::vec::IntoIter<SeafowlPartition>, [closure@src/context.rs:1339:78: 1342:46]>>`
note: future is not `Send` as this value is used across an await
    --> src/context.rs:1365:50
     |
1339 | / ...                       &partitions.into_iter().group_by(|p| {
1340 | | ...                           !partitions_to_filter
1341 | | ...                               .contains(&p.partition_id.unwrap())
1342 | | ...                       })
     | |____________________________- has type `&itertools::GroupBy<bool, std::vec::IntoIter<SeafowlPartition>, [closure@src/context.rs:1339:78: 1342:46]>` which is not `Send`
...
1365 |   ...                           )
     |  ________________________________^
1366 | | ...                           .await?;
     | |____________________________________^ await occurs here, with the value maybe used later
...
1381 |   ...                   }
     |                         - the value is later dropped here
     = note: required for the cast to the object type `dyn futures::Future<Output = std::result::Result<Arc<(dyn ExecutionPlan + 'static)>, datafusion::error::DataFusionError>> + std::marker::Send`
```

Reply:
Oh I see, maybe the groupby has some internal state that isn't safe to be sent across threads. You can avoid the "mut + insert" pattern by using collect() though:

```rust
diff --git a/src/context.rs b/src/context.rs
index 39d8d7e..5fc9bd6 100644
--- a/src/context.rs
+++ b/src/context.rs
@@ -1333,19 +1333,18 @@ impl SeafowlContext for DefaultSeafowlContext {
                                             );
 
                                         // Group adjacent partitions eligible for filtering
-                                        let mut partitions_grouped: Vec<(
+                                        let partitions_grouped: Vec<(
                                             bool,
                                             Vec<SeafowlPartition>,
-                                        )> = Vec::new();
-                                        for (keep, group) in
-                                            &partitions.into_iter().group_by(|p| {
+                                        )> = partitions
+                                            .into_iter()
+                                            .group_by(|p| {
                                                 !partitions_to_filter
                                                     .contains(&p.partition_id.unwrap())
                                             })
-                                        {
-                                            partitions_grouped
-                                                .push((keep, group.collect()))
-                                        }
+                                            .into_iter()
+                                            .map(|(keep, group)| (keep, group.collect()))
+                                            .collect();
 
                                         for (keep, group) in partitions_grouped {
                                             if keep {
```

Reply:
Good point, applying patch.